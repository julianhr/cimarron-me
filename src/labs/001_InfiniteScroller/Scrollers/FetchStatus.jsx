import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'


const Message = styled.div`
  background: #eee;
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
  font-size: 20px;
  padding: 40px;
`

function FetchStatus({ children }) {
  return (
    <Message>
      <div>
        {children}
      </div>
    </Message>
  )
}

FetchStatus.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element,
    PropTypes.string
  ]).isRequired,
}

export default FetchStatus
