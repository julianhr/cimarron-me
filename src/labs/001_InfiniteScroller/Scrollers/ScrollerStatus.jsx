import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'

import { colors } from '~/styles/theme'

const ANIMATION_BUFFER = 30

const Root = styled.div`
  display: flex;
  width: 100%;
  padding: 10px 30px 10px;
`

const Span = styled.span`
  padding: 2px 10px;
`

export class ScrollerStatus extends React.PureComponent {
  static propTypes = {
    entryCount: PropTypes.number,
  }

  static defaultProps = {
    entryCount: 0,
  }

  state = {
    flashState: 'reset',
    fetchCount: 0,
  }

  componentDidUpdate(prevProps, prevState) {
    // reset state when a new scroller method is selected
    if (this.props.entryCount === 0 && this.state.fetchCount > 0) {
      this.setState({ fetchCount: 0, flashState: 'reset' })
    // animate if count has increased
    } else if (prevProps.entryCount < this.props.entryCount) {
      this.setState({ fetchCount: this.state.fetchCount + 1 },
        () => this.canAnimate() && this.animateFlash()
      )
    }
  }

  canAnimate() {
    // don't animate on initial fetch, only when the user
    //  triggers one by scrolling
    return this.state.fetchCount >= 2
  }

  animateFlash() {
    this.setState({ flashState: 'stage' })
    setTimeout(() => { this.setState({ flashState: 'transition' }) }, ANIMATION_BUFFER)
  }

  getSpanStyle() {
    const transDurationMs = 65

    switch (this.state.flashState) {
      case 'reset':
        return {}
      case 'stage':
        return {
          backgroundColor: colors.highlight,
          willChange: 'background-color',
        }
      case 'transition':
        return {
          transition: `background-color ${transDurationMs}ms`,
          backgroundColor: 'transparent',
        }
    }
  }

  renderStatus() {
    const { entryCount } = this.props

    if (entryCount > 0) {
      return (
        <div>
          Records: 
          <Span
            css={this.getSpanStyle()} 
          >
            {entryCount}
          </Span>
        </div>
      )
    }
  }

  render() {
    return (
      <Root>
        {this.renderStatus()}
      </Root>
    )
  }

}

const mapStateToProps = ({ entryCount }) => ({ entryCount })

export default connect(mapStateToProps)(ScrollerStatus)
