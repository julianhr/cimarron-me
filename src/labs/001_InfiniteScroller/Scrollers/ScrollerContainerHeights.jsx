import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'
import { throttle } from 'lodash-es'

import { setIsFetching, setEntryCount } from '../actions/rootActions'
import Card from '../Card'
import FetchStatus from './FetchStatus'


const Root = styled.div`
  height: 100%;
  padding: 10px 10px 0;
  background: #e8e8e8;
  overflow-y: scroll;
`

class ScrollerContainerHeights extends React.PureComponent {
  static propTypes = {
    cardFetcher: PropTypes.func,
    isFetching: PropTypes.bool,
    setEntryCount: PropTypes.func,
    setIsFetching: PropTypes.func,
  }

  state = {
    cards: [],
    fetch: {
      status: 'loading',
      error: null,
    },
  }

  refRoot = React.createRef()

  componentDidMount() {
    this.props.setEntryCount(0)
    this.fetchCards()
  }

  handleOnScroll(event) {
    const { isFetching } = this.props
    const { cards } = this.state

    if (!isFetching && this.canFetchCards(this.refRoot.current) && cards.length < 200) {
      this.fetchCards()
    }
  }
  
  canFetchCards(containerElement) {
    const { scrollHeight, scrollTop, clientHeight } = containerElement
    const buffer = 500
    return scrollHeight - scrollTop < clientHeight + buffer
  }

  async fetchCards() {
    const { cardFetcher, setIsFetching } = this.props
    let data
    setIsFetching(true)

    try {
      data = await cardFetcher(this)
      if (!data) { return }
    } catch (error) {
      this.setState({ fetch: { status: 'error', error: error.toString() } })
      return
    }

    const { cards: currCards } = this.state
    const cards = [...currCards]

    this.appendNewCards(cards, data, currCards.length)
    this.props.setEntryCount(cards.length)
    this.setState({ cards, fetch: null })
    setIsFetching(false)
  }

  appendNewCards(cards, data, lastKey) {
    data.forEach(({ title, image_url: imgUrl, description }, i) => {
      const position = lastKey + i + 1
      const props = { imgUrl, title, description, position }
      cards.push(<Card key={lastKey+i} {...props} /> )
    })
  }

  renderCardResult() {
    const fetchStatus = (this.state.fetch || {}).status

    if (fetchStatus === 'loading') {
      return <FetchStatus>Loading...</FetchStatus>
    } else if (fetchStatus === 'error') {
      return <FetchStatus>{this.state.fetch.error.toString()}</FetchStatus>
    } else {
      return this.state.cards
    }
  }

  render() {
    return (
      <Root
        ref={this.refRoot}
        onScroll={throttle(this.handleOnScroll, 150, { leading: false }).bind(this)}
      >
        {this.renderCardResult()}
      </Root>
    )
  }
}

const mapStateToProps = ({ isFetching }) => ({ isFetching })
const mapDispatchToProps = { setIsFetching, setEntryCount }

export default connect(mapStateToProps, mapDispatchToProps)(ScrollerContainerHeights)
