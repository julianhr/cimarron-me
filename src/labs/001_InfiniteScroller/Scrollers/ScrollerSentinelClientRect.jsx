import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'
import { throttle } from 'lodash-es'

import { setIsFetching, setEntryCount } from '../actions/rootActions'
import Card from '../Card'
import FetchStatus from './FetchStatus'


const Root = styled.div`
  height: 100%;
  padding: 10px 10px 0;
  background: #e8e8e8;
  overflow-y: scroll;
`

export class ScrollerSentinelClientRect extends React.PureComponent {
  static propTypes = {
    cardFetcher: PropTypes.func.isRequired,
    isFetching: PropTypes.bool.isRequired,
    sentinelPosition: PropTypes.number.isRequired,
    setEntryCount: PropTypes.func.isRequired,
    setIsFetching: PropTypes.func.isRequired,
  }

  state = {
    cards: [],
    fetch: {
      status: 'loading',
      error: null,
    },
  }

  refRoot = React.createRef()
  refSentinel = React.createRef()

  componentDidMount() {
    this.props.setEntryCount(0)
    this.getNewCards()
  }

  handleOnScroll() {
    const { isFetching } = this.props
    const { cards } = this.state
    const { top: sentinelTop } = this.refSentinel.current.getBoundingClientRect()
    const { bottom: rootBottom } = this.refRoot.current.getBoundingClientRect()
    const isSentinelVisible = sentinelTop <= rootBottom

    if (!isFetching && isSentinelVisible && cards.length < 200) {
      this.getNewCards()
    }
  }

  async getNewCards() {
    const { cardFetcher, setIsFetching, setEntryCount } = this.props

    try {
      setIsFetching(true)
      const data = await cardFetcher(this)
      this.setNewCards(data)
      setEntryCount(this.state.cards.length)
    } catch (error) {
      this.setState({ fetch: { status: 'error', error: error.toString() } })
    }

    setIsFetching(false)
  }

  setNewCards(data) {
    const { sentinelPosition } = this.props
    const { length: lastKey } = this.state.cards
    const cards = [...this.state.cards]

    data.forEach((datum, i) => {
      const card = this.getCard(datum, i, sentinelPosition, lastKey)
      cards.push(card)
    })

    this.setState({ cards, fetch: null })
  }

  getCard(datum, i, sentinelPosition, lastKey) {
    const { title, image_url: imgUrl, description } = datum
    const position = lastKey + i + 1
    const props = { imgUrl, title, description, position }
    let CardComponent = Card

    if (i + 1 === sentinelPosition) {
      CardComponent = React.forwardRef((props, ref) => <Card forwardedRef={ref} {...props} />)
      // previous ref element will be overwritten by new sentinel
      props.forwardedRef = this.refSentinel
    }

    return <CardComponent key={lastKey + i} {...props} />
  }

  renderResult() {
    const fetchStatus = (this.state.fetch || {}).status

    if (fetchStatus === 'loading') {
      return <FetchStatus>Loading...</FetchStatus>
    } else if (fetchStatus === 'error') {
      return <FetchStatus>{this.state.fetch.error.toString()}</FetchStatus>
    } else {
      return this.state.cards
    }
  }

  render() {
    return (
      <Root
        ref={this.refRoot}
        onScroll={throttle(this.handleOnScroll, 200, { leading: false }).bind(this)}
      >
        {this.renderResult()}
      </Root>
    )
  }
}

const mapStateToProps = ({ sentinelPosition, isFetching }) => (
  { sentinelPosition, isFetching }
)

const mapDispatchToProps = { setIsFetching, setEntryCount }

export default connect(mapStateToProps, mapDispatchToProps)(ScrollerSentinelClientRect)
