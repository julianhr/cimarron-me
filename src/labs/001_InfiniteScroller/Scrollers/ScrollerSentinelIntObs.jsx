import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'

import { setIsFetching, setEntryCount } from '../actions/rootActions'
import Card from '../Card'
import FetchStatus from './FetchStatus'


const Root = styled.div`
  height: 100%;
  padding: 10px 10px 0;
  background: #e8e8e8;
  overflow-y: auto;
`

export class ScrollerSentinelIntObs extends React.PureComponent {
  static propTypes = {
    cardFetcher: PropTypes.func.isRequired,
    isFetching: PropTypes.bool.isRequired,
    sentinelPosition: PropTypes.number.isRequired,
    setEntryCount: PropTypes.func.isRequired,
    setIsFetching: PropTypes.func.isRequired,
  }

  state = {
    isSupported: true,
    cards: [],
    fetch: {
      status: 'loading',
      error: null,
    },
  }

  refSentinel = React.createRef()
  observer = null

  componentDidMount() {
    this.props.setEntryCount(0)

    if (this.isSupported()) {
      this.observer = new IntersectionObserver(this.handleIntObs())
      this.getNewCards()
    } else {
      this.setState({ isSupported: false, })
    }
  }

  componentDidUpdate(prevProps, prevState) {
    const { isSupported, cards } = this.state

    if (isSupported && cards.length > prevState.cards.length) {
      this.observer.observe(this.refSentinel.current)
    }
  }

  isSupported() {
    // Boolean check needed for Node and Jest by extension
    return 'IntersectionObserver' in window && Boolean(IntersectionObserver)
  }

  async getNewCards() {
    const { cardFetcher, setIsFetching, setEntryCount } = this.props

    try {
      setIsFetching(true)
      const data = await cardFetcher(this)
      this.setNewCards(data)
      setEntryCount(this.state.cards.length)
    } catch (error) {
      this.setState({ fetch: { status: 'error', error: error.toString() } })
    }

    setIsFetching(false)
  }

  getCard(datum, i, sentinelPosition, lastKey) {
    const { title, image_url: imgUrl, description } = datum
    const position = lastKey + i + 1
    const props = { imgUrl, title, description, position }
    let CardElement = Card

    if (i + 1 === sentinelPosition) {
      CardElement = React.forwardRef((props, ref) => <Card forwardedRef={ref} {...props} />)
      props.forwardedRef = this.refSentinel
    }

    return <CardElement key={lastKey + i} {...props} />
  }

  setNewCards(data) {
    const { length: lastKey } = this.state.cards
    const { sentinelPosition } = this.props
    const cards = [...this.state.cards]

    // a new sentinel will be set, remove the previous one
    if (this.refSentinel.current) {
      this.observer.unobserve(this.refSentinel.current)
    }

    data.forEach((datum, i) => {
      const card = this.getCard(datum, i, sentinelPosition, lastKey)
      cards.push(card)
    })

    this.setState({ cards, fetch: null })
  }

  handleIntObs() {
    const parent = this

    return (entries) => {
      entries.forEach(entry => {
        const { isFetching } = parent.props
        const { cards } = parent.state

        if (entry.isIntersecting && !isFetching && cards.length < 200) {
          parent.getNewCards()
        }
      })
    }
  }

  renderResult() {
    const fetchStatus = (this.state.fetch || {}).status

    if (!this.state.isSupported) {
      return (
        <FetchStatus>
          <>
            <p>This browser doesn't support the Intersection Observer.</p>
            <a
              href='https://caniuse.com/#feat=intersectionobserver'
              target='_blank'
              rel='noopener noreferrer'
            >
              Check compatibility
            </a> 
          </>
        </FetchStatus>
      )
    } else if (fetchStatus === 'loading') {
      return <FetchStatus>Loading...</FetchStatus>
    } else if (fetchStatus === 'error' && this.state.cards.length === 0) {
      return <FetchStatus>{this.state.fetch.error.toString()}</FetchStatus>
    } else {
      return this.state.cards
    }
  }

  render() {
    return (
      <Root>
        {this.renderResult()}
      </Root>
    )
  }
}

const mapStateToProps = ({ sentinelPosition, isFetching }) => (
  { sentinelPosition, isFetching }
)

const mapDispatchToProps = { setIsFetching, setEntryCount }

export default connect(mapStateToProps, mapDispatchToProps)(ScrollerSentinelIntObs)
