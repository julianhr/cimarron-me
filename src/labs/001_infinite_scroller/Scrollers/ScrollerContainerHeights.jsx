import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'

import { setIsFetching, setEntryCount } from '../actions/rootActions'
import Card from '../Card'
import FetchStatus from './FetchStatus'


const Root = styled.div`
  height: 100%;
  padding: 10px 10px 0;
  background: #e8e8e8;
  overflow-y: scroll;
`

class ScrollerContainerHeights extends React.PureComponent {
  static propTypes = {
    cardFetcher: PropTypes.func,
    isFetching: PropTypes.bool,
    setEntryCount: PropTypes.func,
    setIsFetching: PropTypes.func,
  }

  state = {
    cards: []
  }

  componentDidMount() {
    this.fetchCards()
  }

  handleOnScroll = (event) => {
    const { isFetching } = this.props
    const { cards } = this.state

    if (!isFetching && this.canFetchCards(event.target) && cards.length < 200) {
      this.fetchCards()
    }
  }
  
  canFetchCards(containerElement) {
    const { scrollHeight, scrollTop, clientHeight } = containerElement
    const buffer = 500
    return scrollHeight - scrollTop < clientHeight + buffer
  }

  async fetchCards() {
    const { cardFetcher, setIsFetching } = this.props
    let data
    setIsFetching(true)

    try {
      data = await cardFetcher(this)
      if (!data) { return }
    } catch (error) {
      this.setState({ fetch: { status: 'error', error: error.toString() } })
      return
    }

    const { cards: currCards } = this.state
    const cards = [...currCards]

    this.appendNewCards(cards, data, currCards.length)
    this.props.setEntryCount(cards.length)
    this.setState({ cards, fetch: null })
    setIsFetching(false)
  }

  appendNewCards(cards, data, lastKey) {
    data.forEach(({ title, image_url: imgUrl, description }, i) => {
      const position = lastKey + i + 1
      const props = { imgUrl, title, description, position }
      cards.push(<Card key={lastKey+i} {...props} /> )
    })
  }

  renderCardResult() {
    const fetchStatus = (this.state.fetch || {}).status

    if (fetchStatus === 'loading') {
      return <FetchStatus>Loading...</FetchStatus>
    } else if (fetchStatus === 'error') {
      return <FetchStatus>{this.state.fetch.error.toString()}</FetchStatus>
    } else {
      return this.state.cards
    }
  }

  render() {
    return (
      <Root
        onScroll={this.handleOnScroll}
      >
        {this.renderCardResult()}
      </Root>
    )
  }
}

const mapStateToProps = ({ isFetching }) => ({ isFetching })
const mapDispatchToProps = { setIsFetching, setEntryCount }

export default connect(mapStateToProps, mapDispatchToProps)(ScrollerContainerHeights)
