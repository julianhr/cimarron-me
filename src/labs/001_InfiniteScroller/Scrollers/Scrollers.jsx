import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'

import { urlBuilder } from '~/utils'
import ScrollerStatus from './ScrollerStatus'
import ScrollerSentinelIntObs from './ScrollerSentinelIntObs'
import ScrollerContainerHeights from './ScrollerContainerHeights'
import ScrollerSentinelClientRect from './ScrollerSentinelClientRect'


export const SCROLLERS = {
  intersectionObserver: ScrollerSentinelIntObs,
  containerScrollHeights: ScrollerContainerHeights,
  sentinelClientRect: ScrollerSentinelClientRect,
}

const Root = styled.section`
  width: 100%;
  height: 80vh;
  min-height: 400px;
  max-height: 700px;
  display: flex;
  flex-direction: column;
  padding: 0 30px 20px;
  align-items: center;

  ${props => props.theme.queries.from('md')} {
    max-height: 800px;
  }
`

const Contour = styled.div`
  width: 100%;
  max-width: 600px;
  border: 2px solid #acacac;
  height: 100%;
  overflow: hidden;
`

export class Scrollers extends React.PureComponent {
  static propTypes = {
    recordsPerFetch: PropTypes.number.isRequired,
    scrollerType: PropTypes.string.isRequired,
  }

  async fetchCards() {
    const path = '/infinite-scroller'
    const query = { paragraphs: 1, entries: this.props.recordsPerFetch }
    const url = urlBuilder(path, query)
    const res = await fetch(url)
  
    try {
      if (res.ok) {
        const json = await res.json()
        return json
      } else {
        throw Error('There was an error with the fetch request.')
      }
    } catch (error) {
      console.error('testing error', error)
      throw error
    }
  }

  renderChildren() {
    if (this.props.scrollerType in SCROLLERS) {
      const Element = SCROLLERS[this.props.scrollerType]
      return <Element cardFetcher={this.fetchCards.bind(this)} />
    }
  }

  render() {
    return (
      <Root>
        <ScrollerStatus />
        <Contour>
          {this.renderChildren()}
        </Contour>
      </Root>
    )
  }
}


const mapStateToProps = ({ recordsPerFetch, scrollerType }) => (
  { recordsPerFetch, scrollerType }
)

export default connect(mapStateToProps)(Scrollers)
