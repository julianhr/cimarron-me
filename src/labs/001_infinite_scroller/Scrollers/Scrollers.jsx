import React from 'react'
import PropTypes from 'prop-types'
import styled from '@emotion/styled'
import { connect } from 'react-redux'

import { urlBuilder } from '~/utils'
import ScrollerStatus from './ScrollerStatus'
import ScrollerSentinelIntObs from './ScrollerSentinelIntObs'
import ScrollerContainerHeights from './ScrollerContainerHeights'
import ScrollerSentinelClientRect from './ScrollerSentinelClientRect'


const Root = styled.section`
  width: 100%;
  height: 80vh;
  min-height: 400px;
  max-height: 700px;
  display: flex;
  flex-direction: column;
  padding: 0 30px 20px;
  align-items: center;

  ${props => props.theme.queries.from('md')} {
    max-height: 800px;
  }
`

const Contour = styled.div`
  max-width: 600px;
  border: 2px solid #acacac;
  height: 100%;
  overflow: hidden;
`

function Scrollers({ scrollerType, recordsPerFetch }) {
  const fetchCards = () => {
    const path = '/infinite-scroller'
    const query = { paragraphs: 1, entries: recordsPerFetch }
    const url = urlBuilder(path, query)

    return fetch(url)
      .then(res => {
        if (res.ok) { return res.json() }
        else { throw Error('Fetch error') }
      })
      .catch(error => {
        console.error('Fetch error:', error)
      })
  }

  const renderScroller = () => {
    switch (scrollerType) {
      case 'intersectionObserver':
        return <ScrollerSentinelIntObs cardFetcher={fetchCards} />
      case 'containerScrollHeights':
        return <ScrollerContainerHeights cardFetcher={fetchCards} />
      case 'sentinelClientRect':
        return <ScrollerSentinelClientRect cardFetcher={fetchCards} />
    }
  }

  return (
    <Root>
      <ScrollerStatus />
      <Contour>
        {renderScroller()}
      </Contour>
    </Root>
  )
}

Scrollers.propTypes = {
  scrollerType: PropTypes.string,
  recordsPerFetch: PropTypes.number,
}

const mapStateToProps = ({ recordsPerFetch, scrollerType }) => (
  { scrollerType, recordsPerFetch }
)

export default connect(mapStateToProps)(Scrollers)
