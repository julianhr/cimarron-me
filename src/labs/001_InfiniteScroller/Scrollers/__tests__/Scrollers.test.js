import React from 'react'
import { render, mount, shallow } from 'enzyme'
import { create } from 'react-test-renderer'

import { Scrollers, SCROLLERS } from '../Scrollers'
import RootMock from '../../__tests__/__mocks__/RootMock'


describe('Scrollers', () => {
  const testProps = {
    recordsPerFetch: 5,
    scrollerType: Object.keys(SCROLLERS)[0],
  }

  const renderRooted = (props) => mount(<RootMock><Scrollers {...props} /></RootMock>)

  describe('props', () => {
    it('recordsPerFetch is required', () => {
      const props = {...testProps}
      delete props.recordsPerFetch
      expect(() => renderRooted(props)).toThrow()
    })

    it('scrollerType is required', () => {
      const props = {...testProps}
      delete props.scrollerType
      expect(() => renderRooted(props)).toThrow()
    })
  })

  describe('#fetchCards', () => {
    let { fetch: origFetch } = global

    afterEach(() => {
      global.fetch = origFetch
    })

    it('is async', async () => {
      jest.spyOn(Scrollers.prototype, 'fetchCards')
        .mockImplementation(() => Promise.resolve(true))
      const wrapper = renderRooted(testProps)
      const instance = wrapper.find(Scrollers).instance()
      await expect(instance.fetchCards()).resolves.toBe(true)
    })

    it('returns json if fetch is successful', async () => {
      const jsonFn = () => ({ one: 1, two: 2 })
      const res = { ok: true, json: jsonFn }
      global.fetch = async () => res
      const wrapper = renderRooted(testProps)
      const instance = wrapper.find(Scrollers).instance()
      await expect(instance.fetchCards()).resolves.toEqual(jsonFn())
    })

    it('throw exception if fetch is unsuccessful', async () => {
      global.fetch = async () => ({ ok: false })
      const wrapper = renderRooted(testProps)
      const instance = wrapper.find(Scrollers).instance()
      await expect(instance.fetchCards()).rejects.toThrow()
    })
  })

  it('matches snapshot with all props', () => {
    const wrapper = renderRooted(testProps)
    expect(wrapper).toMatchSnapshot()
    wrapper.unmount()
  })
})
