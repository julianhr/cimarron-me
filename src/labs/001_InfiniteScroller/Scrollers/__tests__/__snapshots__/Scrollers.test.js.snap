// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Scrollers matches snapshot with all props 1`] = `
Array [
  <ForwardRef(render)
    styles={
      Object {
        "map": undefined,
        "name": "1e2xhh9-globalStyles-globalStyles",
        "next": undefined,
        "styles": "body{color:#3D3D3D,lineHeight 1.3em;}p,h1,h2,h3,h4,h5,h6,figure{margin:0;}h1,h2,h3,h4,h5,h6{color:#AB164D;}p{padding:.8em 0;}h1{font-size:44;}h2{line-height:2em;font-size:34;font-weight:500;}h3{font-size:24;font-weight:600;}h4{line-height:1.3em;font-size:18;}h5{font-size:12;}h6{font-size:8;}code{color:#D90000;font-size:14;}label:globalStyles;label:globalStyles;",
      }
    }
  />,
  <Root>
    <section
      className="css-1sjkg4e-Root e1iukibw0"
    >
      <Root>
        <div
          className="css-47ru1c-Root exr8ng20"
        />
      </Root>
      <Contour>
        <div
          className="css-1sfqwic-Contour e1iukibw1"
        >
          <Root>
            <div
              className="css-11qniw9-Root e5woe9i0"
            >
              <Message>
                <div
                  className="css-quardp-Message ehwlun20"
                >
                  <div>
                    <p>
                      This browser doesn't support the Intersection Observer.
                    </p>
                    <a
                      href="https://caniuse.com/#feat=intersectionobserver"
                      rel="noopener noreferrer"
                      target="_blank"
                    >
                      Check compatibility
                    </a>
                  </div>
                </div>
              </Message>
            </div>
          </Root>
        </div>
      </Contour>
    </section>
  </Root>,
]
`;
